<font size="6"> <div class="alert alert-block alert-info"><center>**Primeiro trabalho individual de Otimização Heurística**</center></div></font>

<center><h4> Trabalho realizado por Sebastião Rosalino, nº98437 turma CDB1, Licenciatura de Ciência de Dados</h4></center>

<center><h4> Docente: Anabela Costa </h4></center>

<center><img src="culturas_campos.jpg" width=250px height=150px></center>

<center><img src="logo_iscte.png" width=200px height=200px></center>

#### Importação das bibliotecas necessárias 

In [ ]:
import numpy as np
from pulp import LpMaximize, LpMinimize, LpProblem, LpStatus, lpSum, LpVariable

# **Parte 1**

#### **a) Formule o problema em Programação Linear**

Identificação das variáveis de decisão:

X1 - Milhares de hectares a cultivar com a cultura 1 no próximo ano

X2 - Milhares de hectares a cultivar com a cultura 2 no próximo ano

X3 - Milhares de hectares a cultivar com a cultura 3 no próximo ano

As Funções objetivo são as seguintes:

* Maximizar a quantidade de capital estrangeiro gerado

Max Qt. Cap. Est. = 3000 * X1 + 5000 * X2 + 4000 * X3

Maximiza-se a quantidade de capital estrangeiro pela função que considera que para cada milhar de hectare da cultura 1 obtém-se 3000 euros de capital estrangeiro e assim sucessivamente para as restantes culturas.

* Maximizar o número de cidadãos alimentados

Max N.º Cid. Alim. = 150 * X1 + 75 * X2 + 100 * X3

Maximiza-se a quantidade de cidadãos alimentados pela função que considera que para cada milhar de hectare da cultura 1 se alimentam 150 pessoas e assim sucessivamente para as restantes culturas.

* Maximizar o número de cidadãos contratados para laboração nestas culturas

Max Trabalhadores = 10 * X1 + 15 * X2 + 12 * X3

Maximiza-se a quantidade de trabalhadores pela função que considera que para cada milhar de hectare da cultura 1 se criam 10 empregos e assim sucessivamente para as restantes culturas.


Sujeito a:
    
X1 + X2 + X3 <= 15000 (Milhares de hectares totais disponíveis)
                    
X3 >= 1000 (Mínimo de milhares de hectares a cultivar na cultura 3

X1, X2, X3 >= 0 (Milhares de hectares não negativos a cultivar a todas as culturas) 

#### **b)  Ao avaliar a possibilidade de não conseguir alcançar a maximização dos três fatores, o governo definiu níveis aceitáveis para cada um dos fatores:**

É indicado no enunciado que, ao avaliar a possibilidade de não conseguir alcançar a maximização dos três fatores, o Governo definiu níveis aceitáveis para cada um dos fatores, a saber:

* Capital estrangeiro gerado não inferior a 70 milhões de euros;
* Número de cidadãos alimentados não inferior a 1.75 milhões;
* Aproximadamente 200 mil cidadãos contratados para laboração nas culturas.

A área disponível a cultivar continua a ser os 15000 milhares de hectares de terrenos agrícolas.


#### **b1) Recorrendo à Programação por Metas não Preemptiva, apresente duas propostas de distribuição dos hectares de terreno alocados a cada um dos tipos de cultura.**

A resolução do problema consiste na definição de metas não preemptivas, o que significa que são metas sem prioridade entre si (ou seja, igualmente prioritárias). As metas são as seguintes:

* Meta 1: Capital estrangeiro gerado não muito menos que 70 milhões de euros;
* Meta 2: Número de cidadãos alimentados não muito menos que 1.75 milhões;
* Meta 3: Aproximadamente 200 mil cidadãos contratados para laboração nas culturas.

#### Proposta de resolução pelo método da soma ponderada dos desvios

In [ ]:
model = LpProblem(name="Trabalho_1_parte_1_desvios", sense=LpMinimize)

x = {i: LpVariable(name=f"x{i}", lowBound=0) for i in range(1, 4)}
dm1 = LpVariable(name="dm1", lowBound=0)
dm2 = LpVariable(name="dm2", lowBound=0)
dM3 = LpVariable(name="dM3", lowBound=0)
dm3 = LpVariable(name="dm3", lowBound=0)

model += (3000 * x[1] + 5000 * x[2] + 4000 * x[3] >= 70000000 - dm1, "capital estrangeiro")
model += (150 * x[1] + 75 * x[2] + 100 * x[3] >= 1750000 - dm2, "cidadaos alimentados")
model += (10 * x[1] + 15 * x[2] + 12 * x[3] == 200000 + dM3 - dm3, "cidadaos contratados")
model += (x[1] + x[2] + x[3] <= 15000, "hectares disponiveis")
model += (x[3] >= 1000, 'minimo da cultura 3')

vam = np.array([70000000, 1750000, 200000])

obj_func = (1/vam[0]) * dm1 + (1/vam[1]) * dm2 + (1/vam[2]) * dM3 + (1/vam[2]) * dm3
model += obj_func

model

Trabalho_1_parte_1_desvios:
MINIMIZE
5e-06*dM3 + 1.4285714285714286e-08*dm1 + 5.714285714285714e-07*dm2 + 5e-06*dm3 + 0.0
SUBJECT TO
capital_estrangeiro: dm1 + 3000 x1 + 5000 x2 + 4000 x3 >= 70000000

cidadaos_alimentados: dm2 + 150 x1 + 75 x2 + 100 x3 >= 1750000

cidadaos_contratados: - dM3 + dm3 + 10 x1 + 15 x2 + 12 x3 = 200000

hectares_disponiveis: x1 + x2 + x3 <= 15000

minimo_da_cultura_3: x3 >= 1000

VARIABLES
dM3 Continuous
dm1 Continuous
dm2 Continuous
dm3 Continuous
x1 Continuous
x2 Continuous
x3 Continuous

In [ ]:
# Resolução do modelo

status = model.solve()

print(f"objective: {model.objective.value()}")

for var in x.values():
    print(f"{var.name}: {var.value()}")
print(f"{dm1}: {dm1.value()}")
print(f"{dm2}: {dm2.value()}")
print(f"{dM3}: {dM3.value()}")
print(f"{dm3}: {dm3.value()}")


for name, constraint in model.constraints.items():
    print (f"{name}: {constraint.value()}")

objective: 0.22285714285714286
x1: 4400.0
x2: 9600.0
x3: 1000.0
dm1: 4800000.0
dm2: 270000.0
dM3: 0.0
dm3: 0.0
capital_estrangeiro: 0.0
cidadaos_alimentados: 0.0
cidadaos_contratados: 0.0
hectares_disponiveis: 0.0
minimo_da_cultura_3: 0.0


Podemos verificar que, aplicando o método da soma ponderada dos desvios, os resultados das variáveis de decisão são os seguintes (alocação ótima):

X1 - Milhares de hectares a cultivar com a cultura 1 = 4400

X2 - Milhares de hectares a cultivar com a cultura 2 = 9600

X3 - Milhares de hectares a cultivar com a cultura 3 = 1000

A totalidade do terreno disponível (15000 milhares de hectares) está a ser utilizada.
Em termos dos objetivos propostos (metas), os resultados obtidos indicam-nos que:

Relativamente à Meta 1, o desvio é de 4800000 euros de capital estrangeiro gerado, ou seja, menos 6,86% face ao objetivo dos 70000000 de euros;

Relativamente à Meta 2, o desvio é de 270000 cidadãos alimentados, ou seja, menos 15,43% face ao objetivo dos 1750000 milhões de cidadãos;

A Meta 3, aproximadamente 200000 cidadãos contratados para laboração nas culturas, é integralmente cumprida, na medida em que é exatamente esse o número de cidadão a laborar nas 3 culturas. Ou seja, os desvios (dM3 e dm3) são zero.

O valor ótimo (objetive) situa-se em 0.22286, o que à partida, por estar relativamente afastado de zero, deixa algumas dúvidas sobre a qualidade e validade deste indicador. Por isso, o mais importante é interpretar a solução ótima.

Em termos gerais, podemos concluir que a solução é equilibrada, desde logo porque se consegue utilizar todo o terreno disponível. Mas também porque as metas são relativamente cumpridas, no sentido em que os desvios verificados não são excessivos. 

Para a meta 1, o desvio é ligeiramente superior a 5%, o que pode ser considerado bastante positivo. Para a meta 2 o desvio é de cerca de 15%, que pode ser considerado um pouco excessivo, mas não em demasia num contexto geral. A meta 3 é integralmente cumprida, sem qualquer desvio.

Perante estes desvios, a solução parece globalmente boa, resultando na obtenção de:

* 65200000 de capital estrangeiro gerado;
* 1480000 cidadãos alimentados;
* 200000 cidadãos contratados.

#### Proposta de resolução pelo método MiniMax

In [ ]:
model = LpProblem(name="Trabalho_1_parte_1_minimax", sense=LpMinimize)
peso = [1, 1, 1, 1,]
vam = [70000000, 1750000, 200000]

x = {i: LpVariable(name=f"x{i}", lowBound=0) for i in range(1, 4)}
dm1 = LpVariable(name="dm1", lowBound=0)
dm2 = LpVariable(name="dm2", lowBound=0)
dM3 = LpVariable(name="dM3", lowBound=0)
dm3 = LpVariable(name="dm3", lowBound=0)
Q = LpVariable(name="Q", lowBound=0)

model += (3000 * x[1] + 5000 * x[2] + 4000 * x[3] >= 70000000 - dm1, "capital estrangeiro")
model += (150 * x[1] + 75 * x[2] + 100 * x[3] >= 1750000 - dm2, "cidadaos alimentados")
model += (10 * x[1] + 15 * x[2] + 12 * x[3] == 200000 + dM3 - dm3, "cidadaos contratados")
model += (x[1] + x[2] + x[3] <= 15000, "hectares disponiveis")
model += (x[3] >= 1000, 'minimo da cultura 3')
model += ((peso[0]/vam[0]) * dm1 - Q <= 0, "desvio_inf_capital_estrangeiro")
model += ((peso[1]/vam[1]) * dm2 - Q <= 0, "desvio_inf_cidadaos_alimentados")
model += ((peso[2]/vam[2]) * dM3 - Q <= 0, "desvio_sup_cidadaos_contratados")
model += ((peso[3]/vam[2]) * dm3 - Q <= 0, "desvio_inf_cidadaos_contratados")

obj_func = Q
model += obj_func

model

Trabalho_1_parte_1_minimax:
MINIMIZE
1*Q + 0
SUBJECT TO
capital_estrangeiro: dm1 + 3000 x1 + 5000 x2 + 4000 x3 >= 70000000

cidadaos_alimentados: dm2 + 150 x1 + 75 x2 + 100 x3 >= 1750000

cidadaos_contratados: - dM3 + dm3 + 10 x1 + 15 x2 + 12 x3 = 200000

hectares_disponiveis: x1 + x2 + x3 <= 15000

minimo_da_cultura_3: x3 >= 1000

desvio_inf_capital_estrangeiro: - Q + 1.42857142857e-08 dm1 <= 0

desvio_inf_cidadaos_alimentados: - Q + 5.71428571429e-07 dm2 <= 0

desvio_sup_cidadaos_contratados: - Q + 5e-06 dM3 <= 0

desvio_inf_cidadaos_contratados: - Q + 5e-06 dm3 <= 0

VARIABLES
Q Continuous
dM3 Continuous
dm1 Continuous
dm2 Continuous
dm3 Continuous
x1 Continuous
x2 Continuous
x3 Continuous

In [ ]:
# Resolução do modelo

status = model.solve()

print(f"objective: {model.objective.value()}")

for var in x.values():
    print(f"{var.name}: {var.value()}")
print(f"{dm1}: {dm1.value()}")
print(f"{dm2}: {dm2.value()}")
print(f"{dM3}: {dM3.value()}")
print(f"{dm3}: {dm3.value()}")


for name, constraint in model.constraints.items():
    print (f"{name}: {constraint.value()}")

objective: 0.10285714
x1: 5600.0
x2: 8400.0
x3: 1000.0
dm1: 7200000.0
dm2: 180000.0
dM3: 0.0
dm3: 6000.0
capital_estrangeiro: 0.0
cidadaos_alimentados: 0.0
cidadaos_contratados: 0.0
hectares_disponiveis: 0.0
minimo_da_cultura_3: 0.0
desvio_inf_capital_estrangeiro: 2.8571428556389833e-09
desvio_inf_cidadaos_alimentados: 2.8571428556389833e-09
desvio_sup_cidadaos_contratados: -0.10285714
desvio_inf_cidadaos_contratados: -0.07285714


Podemos verificar que, aplicando o método do MiniMax, os resultados das variáveis de decisão são os seguintes (alocação ótima):

X1 - Milhares de hectares a cultivar com a cultura 1 = 5600

X2 - Milhares de hectares a cultivar com a cultura 2 = 8400

X3 - Milhares de hectares a cultivar com a cultura 3 = 1000

A totalidade do terreno disponível (15000 milhares de hectares) está a ser utilizada.

Em termos dos objetivos propostos (metas), os resultados obtidos indicam-nos que:

Relativamente à Meta 1, o desvio é de 7200000 euros de capital estrangeiro gerado, ou seja, cerca de 10.3% face ao objetivo dos 70000000 de euros;

Relativamente à Meta 2, o desvio é de 180000 cidadãos alimentados, ou seja, cerca de 10,3% face ao objetivo dos 1750000 milhões de cidadãos;

Relativamente à Meta 3, o desvio é 6000 cidadãos contratados, por defeito, o que representa um gap de cerca de cerca de 3% face ao objetivo de 200000 cidadãos contratados para laboração nas culturas.

O valor ótimo (objective) foi 0.102857, correspondendo ao desvio máximo ocorrido. Atendendo aos valores dos desvios das restrições dm1 e dm2, conclui-se que este desvio ocorreu nas Metas 1 e 2.

Em termos gerais, podemos concluir que a solução é equilibrada, desde logo porque se consegue utilizar todo o terreno disponível. Mas também porque as metas são cumpridas com desvios máximos na ordem dos 10%. Para a meta 1, o desvio é 10,3%, o que pode ser considerado aceitável. Para a meta 2 o desvio é igualmente de 10,3%. Por seu lado, a meta 3 apresenta um desvio de apenas 3%.

Perante estes desvios, a solução parece globalmente boa, resultando na obtenção de:

* 62800000 de capital estrangeiro gerado;
* 1570000 cidadãos alimentados;
* 194000 cidadãos contratados.

#### **b2) Com o objetivo de gerar soluções alternativas, o governo decidiu que a primeira prioriodade deve ser dada ao número de cidadãos alimentados, a segunda prioridade ao capital estrangeiro e a terceira prioridade ao emprego local. Apresente uma solução que respeite as prioridades definidas.** 

In [ ]:
model = LpProblem(name="Trabalho_1_parte_1_primeira_prioridade", sense=LpMinimize)

x = {i: LpVariable(name=f"x{i}", lowBound=0) for i in range(1, 4)}
dm2 = LpVariable(name="dm2", lowBound=0)

model += (3000 * x[1] + 5000 * x[2] + 4000 * x[3] >= 70000000 - dm1, "capital estrangeiro")
model += (150 * x[1] + 75 * x[2] + 100 * x[3] >= 1750000 - dm2, "cidadaos alimentados")
model += (10 * x[1] + 15 * x[2] + 12 * x[3] == 200000 + dM3 - dm3, "cidadaos contratados")
model += (x[1] + x[2] + x[3] <= 15000, "hectares disponiveis")
model += (x[3] >= 1000, 'minimo da cultura 3')

obj_func = dm2
model += obj_func

model

Trabalho_1_parte_1_primeira_prioridade:
MINIMIZE
1*dm2 + 0
SUBJECT TO
capital_estrangeiro: dm1 + 3000 x1 + 5000 x2 + 4000 x3 >= 70000000

cidadaos_alimentados: dm2 + 150 x1 + 75 x2 + 100 x3 >= 1750000

cidadaos_contratados: - dM3 + dm3 + 10 x1 + 15 x2 + 12 x3 = 200000

hectares_disponiveis: x1 + x2 + x3 <= 15000

minimo_da_cultura_3: x3 >= 1000

VARIABLES
dM3 Continuous
dm1 Continuous
dm2 Continuous
dm3 Continuous
x1 Continuous
x2 Continuous
x3 Continuous

In [ ]:
# Resolução do modelo

status = model.solve()

print(f"objective: {model.objective.value()}")

for var in x.values():
    print(f"{var.name}: {var.value()}")
print(f"{dm1}: {dm1.value()}")
print(f"{dm2}: {dm2.value()}")
print(f"{dM3}: {dM3.value()}")
print(f"{dm3}: {dm3.value()}")


for name, constraint in model.constraints.items():
    print (f"{name}: {constraint.value()}")

objective: 0.0
x1: 14000.0
x2: 0.0
x3: 1000.0
dm1: 24000000.0
dm2: 0.0
dM3: 0.0
dm3: 48000.0
capital_estrangeiro: 0.0
cidadaos_alimentados: 450000.0
cidadaos_contratados: 0.0
hectares_disponiveis: 0.0
minimo_da_cultura_3: 0.0


Após construção e resolução do modelo em que se considera apenas a função objetivo de minimizar o desvio da Meta Primária e as restrições hard obteve-se a seguinte solução:

X1 - Milhares de hectares a cultivar com a cultura 1 = 14000

X2 - Milhares de hectares a cultivar com a cultura 2 = 0

X3 - Milhares de hectares a cultivar com a cultura 3 = 1000

O valor ótimo (objective) foi 0, conclui-se assim que a primeira prioridade (número de cidadãos alimentados) foi atingida, obtendo-se a solução de cultivar 14000 milhares de hectares da cultura 1, 0 hectares da cultura 2 e 1000 milhares de hectares da cultura 3. 
É, portanto, possível satisfazer a prioridade de nível 1. É de salientar que nesta solução foram utilizados os 15000 milhares de hectares disponíveis.   

Sabendo que é possível satisfazer a prioridade de nível 1, construiu-se e resolveu-se um novo modelo, cuja função objetivo é minimizar o desvio da Meta Secundária e que nas restrições obrigue a que se cumpra as restrições hard, a restrição do nível 1 e as restrições do nível 2.

In [ ]:
model = LpProblem(name="Trabalho_1_parte_1_segunda_prioridade", sense=LpMinimize)

x = {i: LpVariable(name=f"x{i}", lowBound=0) for i in range(1, 4)}
dm1 = LpVariable(name="dm1", lowBound=0)

model += (3000 * x[1] + 5000 * x[2] + 4000 * x[3] >= 70000000 - dm1, "capital estrangeiro")
model += (150 * x[1] + 75 * x[2] + 100 * x[3] >= 1750000, "cidadaos alimentados")
model += (10 * x[1] + 15 * x[2] + 12 * x[3] == 200000 + dM3 - dm3, "cidadaos contratados")
model += (x[1] + x[2] + x[3] <= 15000, "hectares disponiveis")
model += (x[3] >= 1000, 'minimo da cultura 3')

obj_func = dm1
model += obj_func

model

Trabalho_1_parte_1_segunda_prioridade:
MINIMIZE
1*dm1 + 0
SUBJECT TO
capital_estrangeiro: dm1 + 3000 x1 + 5000 x2 + 4000 x3 >= 70000000

cidadaos_alimentados: 150 x1 + 75 x2 + 100 x3 >= 1750000

cidadaos_contratados: - dM3 + dm3 + 10 x1 + 15 x2 + 12 x3 = 200000

hectares_disponiveis: x1 + x2 + x3 <= 15000

minimo_da_cultura_3: x3 >= 1000

VARIABLES
dM3 Continuous
dm1 Continuous
dm3 Continuous
x1 Continuous
x2 Continuous
x3 Continuous

In [ ]:
# Resolução do modelo

status = model.solve()

print(f"objective: {model.objective.value()}")

for var in x.values():
    print(f"{var.name}: {var.value()}")
print(f"{dm1}: {dm1.value()}")
print(f"{dM3}: {dM3.value()}")
print(f"{dm3}: {dm3.value()}")


for name, constraint in model.constraints.items():
    print (f"{name}: {constraint.value()}")

objective: 12000000.0
x1: 8000.0
x2: 6000.0
x3: 1000.0
dm1: 12000000.0
dM3: 0.0
dm3: 18000.0
capital_estrangeiro: 0.0
cidadaos_alimentados: 0.0
cidadaos_contratados: 0.0
hectares_disponiveis: 0.0
minimo_da_cultura_3: 0.0


Após construção e resolução do modelo em que se considera apenas a função objetivo de minimizar o desvio da Meta Secundária e as restrições hard obteve-se a seguinte solução:

X1 - Milhares de hectares a cultivar com a cultura 1 = 8000

X2 - Milhares de hectares a cultivar com a cultura 2 = 6000

X3 - Milhares de hectares a cultivar com a cultura 3 = 1000

A totalidade do terreno disponível (15000 milhares de hectares) está a ser utilizada.

O valor ótimo (objective) foi 12000000, ou seja, não foi possível cumprir a segunda prioridade (capital estrangeiro gerado não inferior a 70000000 de euros). A solução obtida gera 58000000 de euros de capital estrangeiro, falhando assim a prioridade secundária por 12000000 de euros.

Dada a impossibilidade de cumprir a Meta Secundária no valor inicialmente definido, procedeu-se à redefinição desta meta para o melhor valor gerado (58000000 de euros).

Assumida a nova Meta Secundária, procedeu-se à formulação do problema de modo a que se procure cumprir a Meta Terciária (emprego local).

#### Solução Ótima Final:

In [ ]:
model = LpProblem(name="Trabalho_1_parte_1_terceira_prioridade", sense=LpMinimize)

x = {i: LpVariable(name=f"x{i}", lowBound=0) for i in range(1, 4)}
dm1 = LpVariable(name="dm1", lowBound=0)

model += (3000 * x[1] + 5000 * x[2] + 4000 * x[3] >= 70000000 - 12000000, "capital estrangeiro")
model += (150 * x[1] + 75 * x[2] + 100 * x[3] >= 1750000, "cidadaos alimentados")
model += (10 * x[1] + 15 * x[2] + 12 * x[3] == 200000 + dM3 - dm3, "cidadaos contratados")
model += (x[1] + x[2] + x[3] <= 15000, "hectares disponiveis")
model += (x[3] >= 1000, 'minimo da cultura 3')

obj_func = dM3 + dm3
model += obj_func

model

Trabalho_1_parte_1_terceira_prioridade:
MINIMIZE
1*dM3 + 1*dm3 + 0
SUBJECT TO
capital_estrangeiro: 3000 x1 + 5000 x2 + 4000 x3 >= 58000000

cidadaos_alimentados: 150 x1 + 75 x2 + 100 x3 >= 1750000

cidadaos_contratados: - dM3 + dm3 + 10 x1 + 15 x2 + 12 x3 = 200000

hectares_disponiveis: x1 + x2 + x3 <= 15000

minimo_da_cultura_3: x3 >= 1000

VARIABLES
dM3 Continuous
dm3 Continuous
x1 Continuous
x2 Continuous
x3 Continuous

In [ ]:
# Resolução do modelo

status = model.solve()

print(f"objective: {model.objective.value()}")

for var in x.values():
    print(f"{var.name}: {var.value()}")
print(f"{dM3}: {dM3.value()}")
print(f"{dm3}: {dm3.value()}")


for name, constraint in model.constraints.items():
    print (f"{name}: {constraint.value()}")

objective: 18000.0
x1: 8000.0
x2: 6000.0
x3: 1000.0
dM3: 0.0
dm3: 18000.0
capital_estrangeiro: 0.0
cidadaos_alimentados: 0.0
cidadaos_contratados: 0.0
hectares_disponiveis: 0.0
minimo_da_cultura_3: 0.0


Os resultados das variáveis de decisão são os seguintes (alocação ótima):
    
X1 - Milhares de hectares a cultivar com a cultura 1 = 8000

X2 - Milhares de hectares a cultivar com a cultura 2 = 6000

X3 - Milhares de hectares a cultivar com a cultura 3 = 1000

A totalidade do terreno disponível (15000 milhares de hectares) está a ser utilizada.

Em termos dos objetivos propostos (metas), os resultados obtidos indicam-nos que:

* A meta prioritária foi cumprida: são alimentados 1750000 cidadãos;
* A meta secundária não foi integralmente cumprida na sua formulação inicial, tendo sido gerado 58000000 de capital estrangeiro;
* A meta terciária também não foi, como esperado, cumprida, tendo sido contratados 182000 cidadãos.

## **Parte 2**

#### **a) Quantas versões existem do sistema de portal? E do sistema de rastreio? Justifique as suas respostas.**

### Sistema 1 (Sistema de portal)

Considerou-se que a combinação entre “custo de aquisição”, “custo de manutenção” e “taxa de falsos alarmes” corresponde a uma versão do sistema 1, pelo que se conclui que existem as 9 versões, abaixo descritas:

<table>
    <tr>
        <th> Custo de aquisição em euros </th>
        <th> Custo de manutenção em euros </th>
        <th> Taxa de falso alarme em %</th>
    </tr>
    <tr>
        <td> 90000 </td>
        <td> 15000 </td>
        <td> 10 </td>
     </tr>
     <tr>   
        <td> 105000 </td>
        <td> 16500 </td>
        <td> 9 </td>
    </tr>
    <tr>
        <td> 120000 </td>
        <td> 18000 </td>
        <td> 8 </td>
    </tr>
    <tr>
        <td> 135000 </td>
        <td> 19500 </td>
        <td> 7 </td>
    </tr>
    <tr>  
        <td> 150000 </td>
        <td> 21000 </td>
        <td> 6 </td>
    </tr>
    <tr>
        <td> 165000 </td>
        <td> 22500 </td>
        <td> 5 </td>
    </tr>
    <tr>
        <td> 180000 </td>
        <td> 24000 </td>
        <td> 4 </td>
    </tr>
    <tr>
        <td> 195000 </td>
        <td> 25500 </td>
        <td> 3 </td>
    </tr>
    <tr>
        <td> 210000 </td>
        <td> 27000 </td>
        <td> 2 </td>
    </tr>
</table>

### Sistema 2 (Sistema de rastreio)

Considerou-se que a combinação entre “custo de aquisição”, “custo de manutenção” e “taxa de falsos alarmes” corresponde a uma versão do sistema 2, pelo que se conclui que existem 4 versões, abaixo descritas:

<table>
    <tr>
        <th> Custo de aquisição em euros </th>
        <th> Custo de manutenção em euros </th>
        <th> Taxa de falso alarme em %</th>
    </tr>
    <tr>
        <td> 60000 </td>
        <td> 9000 </td>
        <td> 6 </td>
     </tr>
     <tr>   
        <td> 90000 </td>
        <td> 10200 </td>
        <td> 5 </td>
    </tr>
    <tr>
        <td> 120000 </td>
        <td> 11400 </td>
        <td> 4 </td>
    </tr>
    <tr>
        <td> 150000 </td>
        <td> 12600 </td>
        <td> 3 </td>
    </tr>
</table>

#### **b) No âmbito da Otimização Multiobjetivo, indique como procederia para resolver este problema (mas não resolva)**

A otimização multiobjectivo consiste, genericamente, na resolução problemas em que se pretende alcançar de forma ótima um conjunto de objetivos que, geralmente, são conflituantes. Ou seja, a resolução de um problema multiobjectivo consiste na determinação de soluções de compromisso entre os objetivos em presença.

Para o presente caso, a formulação do problema no âmbito da Otimização Multiobjectivo obedeceria ao seguinte:
Identificação das variáveis de decisão:

X1 - Versão do sistema a 1 utilizar

X2 - Versão do sistema a 2 utilizar

Funções objetivo:

* Objetivo 1: Minimizar a taxa de falsos alarmes do sistema 1

Min Falsos Alarmes = 10 - (X1 - 1)
                
Partindo de uma taxa de 10% de falsos alarmes para a versão mais básica do sistema 1, minimizam-se os falsos alarmes reduzindo 1% por cada versão mais evoluída que se vai introduzindo até à versão 9 em que a taxa de falsos alarmes será de 2%.

* Objetivo 2: Minimizar a taxa de falsos alarmes do sistema 2

Min Falsos Alarmes = 6 - (X2 - 1)
              
Partindo de uma taxa de 6% de falsos alarmes para a versão mais básica do sistema 2, minimizam-se os falsos alarmes reduzindo 1% por cada versão mais evoluída que se vai introduzindo até à versão 4 em que a taxa de falsos alarmes será de 3%.

Sujeito às restrições:

s.a.

X1, X2 são variáveis inteiras

1 <= X1 <= 9 (Versão do sistema 1 a escolher tendo em conta que existem apenas 9)

1 <= X2 <= 4 (Versão do sistema 2 a escolher tendo em conta que existem apenas 4)

90000 + 15000 * (X1 - 1) + 60000 + 30000 * (X2 - 1) <= 250000

Restrição orçamental 1: Despesa total de aquisição, para os dois sistemas, não superior a 250000 euros

15000 + (1500 * (X1 - 1)) + 9000 + (1200 * (X2 - 1)) <= 30000

Restrição orçamental 2: Despesa de manutenção anual, para os dois sistemas, não superior a 30000 euros

<b> Forma alternativa de apresentação do problema (juntando as duas funções objetivo): </b>

Identificação das variáveis de decisão:

X1 - Versão do sistema a 1 utilizar

X2 - Versão do sistema a 2 utilizar

Função objetivo:

Objetivo: Minimizar a taxa de falsos alarmes dos sistemas 1 e 2

Min Falsos Alarmes = 10 - (X1 - 1) + 6 - (X2 - 1)
          
s.a.

X1, X2 são variáveis inteiras

1 <= X1 <= 9 (Versão do sistema 1 a escolher tendo em conta que existem apenas 9)

1 <= X2 <= 4 (Versão do sistema 2 a escolher tendo em conta que existem apenas 4)

90000 + 15000 * (X1 - 1) + 60000 + 30000 * (X2 - 1) <= 250000

Restrição orçamental 1: Despesa total de aquisição, para os dois sistemas, não superior a 250000 euros

15000 + 1500 * (X1 - 1) + 9000 + 1200 * (X2 - 1) <= 30000

Restrição orçamental 2: Despesa de manutenção anual, para os dois sistemas, não superior a 30000 euros